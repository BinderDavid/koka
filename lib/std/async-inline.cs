/*---------------------------------------------------------------------------
  Copyright 2017 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/
using System.Threading;

static class _Async
{
  public class ThreadTimer {
    private Primitive.EventloopEntry entry;
    private Timer timer;

    public ThreadTimer( Fun0<Unit> cb, int ms ) {
      entry = Primitive.GetEventloopEntry();
      if (ms <= 0) {
        timer = null;
        entry.Post(() => { cb.Apply(); });
      }
      else {
        timer = new Timer( (object state0) => { 
                  if (entry != null) entry.Post(() => { cb.Apply(); }); 
                }, null, ms, Timeout.Infinite );
      }
    }

    public void Close() {
      if (entry != null) entry.Close();
      if (timer != null) timer.Dispose();
      entry = null;
      timer = null;
    }
  }

  public static ThreadTimer SetTimeout( Fun0<Unit> cb, int ms ) {
    return new ThreadTimer(cb,ms);
  }

  public static void ClearTimeout( object obj ) {
    if (obj == null) return;
    ThreadTimer t = (ThreadTimer)obj;
    t.Close();
  }
}
