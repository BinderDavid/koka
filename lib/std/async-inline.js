/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

var effCancel = "std/async/cancelable";

function _create_cancelable( caction ) {
  // identity operations
  function ret( loc, x, _k ) { return _k(x); };
  function ops(_l1, _l2, _cont, _op, _resume , _k ) { return _resume(_l1, _l2, _cont, _unit_, _k);  };
  var handle = $std_core._makeHandler1( effCancel, ret, ops );
  var loc = { canceled:false, cancelers:[] };
  function action(_k) { return caction( function(){ return _cancel(loc); }, _k); }
  return function(_k){ return handle( loc, action, _k ); } 
}

function throw_canceled() {
  $std_core.throw_1("Canceled",Cancel);
}

function _cancel( loc ) {
  loc.canceled = true;
  var cs = loc.cancelers; loc.cancelers = [];
  cs.forEach( function(cb) { 
    if (typeof cb.cancel === "function") { 
      cb.cancel();
      setTimeout( function() {
        $std_core._restore_handler_context(cb.hctx, function(_) {
          throw_canceled(); // throw a Cancel exception in that context
        });
      }, 0);
    }
  });
  //return $std_core._unit_;
}

function _check_canceled( hctx ) {
  if (hctx==null) hctx = $std_core._save_handler_context();
  var top = hctx.htop;
  while (top >= 0) {
    var h = hctx.hstack[top];
    if (h.optag===effCancel && h.loc1.canceled) throw_canceled();
    top--;
  }
  //return $std_core._unit_;
}

function _register_cancel( cb, f ) {
  var hctx = $std_core._save_handler_context();
  var top = hctx.htop;
  while (top >= 0 && hctx.hstack[top].optag !== effCancel) { top--; }
  if (top >= 0) {
    //console.log("set cancel: " + cb + ", f: " + f);
    // only register if a cancelation frame exists
    cb.loc = hctx.hstack[top].loc1
    cb.loc.cancelers.push(cb);
    cb.cancel = f;
  }
  //return $std_core._unit_;
}

function _unregister_cancel( cb ) {
  if (typeof cb.cancel === "function") {
    cb.loc.cancelers = cb.loc.cancelers.filter( function(f) { return (f !== cb); } );
  }
  //return $std_core._unit_;
}



// --------------------------------------------------------
// Support for asynchronous actions with effect handlers
// --------------------------------------------------------

var _syncNesting = 0;
var _syncBlocked = null;

function $asyncStartSynchronous() {
  _syncNesting++;
  if (_syncBlocked===null) _syncBlocked = [];
}

function $asyncEndSynchronous() {
  if (_syncNesting===0) return;
  _syncNesting--; 
  if (_syncNesting===0) {
    // now invoke all queued callbacks; we schedule all through a timeout
    _syncBlocked.forEach( function(entry) {
      var cb = function() { return entry.fun.apply(entry.fun,entry.args); };
      setTimeout(cb,0);
    });
    _syncBlocked = null;
  }
}

// Create a callback wrapper around 'cb'. 
// If 'haserr' is true, the first parameter of the new call back is considered an error object
// and thrown automatically if not null. If it is null, it is not passed to the callback `cb`.
// If `nodup` is true, multiple call backs that are queued due to synchronized blocks are not duplicated
// and only the last call back is queued (for timers, mouse events etc).
function $asyncCallback( cb, haserr, nodup ) {
  // save the handler context
  var _hctx = $std_core._save_handler_context();
  // check for cancelation
  _check_canceled(_hctx);
  // remember if this callback itself was constructed inside a sync block
  var _in_synchronous = (_syncNesting > 0); 
  // and return a wrapped callback that restores the context
  var self = null;
  self = (function() {
    var args = Array.prototype.slice.call(arguments);
    // unregister cancelation
    _unregister_cancel(self);
    // check if we return while a 'synchronous' block was started
    if (!self.in_synchronous && _syncNesting > 0) { // if so, put ourselves on the queue
      if (nodup) {
        // update in-place possible previous occurrence of this event if 'nodup' was set (for timers or mouse events etc.)
        for (var i = 0; i < _syncBlocked.length; i++) {
          if (_syncBlocked[i].fun == self) {
            _syncBlocked[i].args = args;
            return; // updated in place, we are done.
          }
        }        
      }
      // and push ourselves on the queue to be called later
      _syncBlocked.push( { fun: self, args: args } );
      return;
    }
    else {
      // restore the handler context and invoke the callback
      $std_core._restore_handler_context( self.hctx, function(_) {
        if (haserr) {
          if (args[0] != null) throw args[0];
          args.shift();         
        }
        return cb.apply(cb,args);
      }, $std_core._unit_ );
    }
  });
  self.hctx = _hctx;
  self.in_synchronous = _in_synchronous;
  self.loc = null;
  self.cancel = null;
  self.canceled = false;
  return self;
}

function $asyncFork( action ) {
  setTimeout( _asyncCallback( function() {
    action($std_core.async_yield);
  }), 0 );  
}

function $asyncAll( actions, k ) {
  var n       = actions.length;
  var results = new Array(n);
  var exn;
  if (n<=0) return k(results);
  
  // all concurrent actions end with this continuation
  // only when all are finished, we resume the passed in continuation
  function allk(i,x,f) {
    results[i] = x;
    n--;
    if (n <= 0) {
      // if an exception was set, raise it now (or keep raising on multiple resume!)
      if (exn!==undefined) throw exn;
      // in case a branch exits more than once (due to amb for example)
      // note: this should never happen as multiple resumes can only happen after
      // at least one resumption but that cannot be past an asyncAll..
      for(i = 0; i < results.length; i++) {
        if (results[i] === undefined) return f(x); 
      }
      return k(results);
    }
    else {
      return f(x);
    }
  }

  for( var i = 0; i < n; i++) {
    (function(j){ 
      var cb = $asyncCallback( function(_) {
        return $std_core.$catch( function(k1) {
            return actions[j](k1);
          },
          function(_exn,_k2) {
            if (exn===undefined) exn = _exn;
            return allk(j, undefined, $std_core.async_yield);
          },
          function(x) { 
            return allk(j,x,$std_core.Result); 
          });
        }); 
      // by using setTimeout we can schedule all continuations for execution
      setTimeout(cb,0); 
    })(i);  // capture i by value
  }
  return $std_core.async_yield();
}
