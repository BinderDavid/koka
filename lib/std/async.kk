/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

// Asynchronous primitives
module std/async

import std/num/int32
import std/time/duration

extern include {
  js file "async-inline.js"
}

type extend exception-info {
  Cancel
}

public fun cancelable( action : (cancel:() -> exn ()) -> <async,exn|e> a ) : <async,exn|e> maybe<a> { 
  val caction = create-cancelable-prim(action)
  catch { Just(caction()) } fun(exn) {
    match(exn.info) {
      Cancel -> {
        check-canceled() // will re-throw if action is not the one that was canceled
        Nothing
      }
      _ -> throw(exn)
    }
  }
}

extern create-cancelable-prim( action : (cancel:() -> exn ()) -> <async,exn|e> a ) : (() -> <async,exn|e> a) {
  js "_create_cancelable"
}

extern check-canceled() : exn () {
  js "_check_canceled"
}

public extern unsafe-register-cancel( cb : a -> <async|e> (), cancel : () -> exn () ) : exn () {
  js "_register_cancel"
}

public fun timeout( secs : duration, action : () -> <async,exn|e> a ) : <async,exn|e> maybe<a> {
  cancelable fun(cancel) {
    val h = set-timeout( fun(_){ cancel(); }, secs.milli-seconds.int32 )
    on-exit({ cancel-timeout(h) }, action)
  }
}

// ----------------------------------------------------------------------------
// Asynchronous operations are supported through the `:async` effect.
// ----------------------------------------------------------------------------

public fun async-fork( action1 : () -> <async|e> a, action2 : () -> <async|e> ()) : <async|e> a {
  val xs = async-all( [{ Just(action1()) }, { action2(); Nothing }] )
  unsafe-noexn{ xs.head.unjust.unjust }
}

// Execute `actions` asynchronously, i.e. interleaved around asynchronous operations.
// When all operations have terminated (or raised an exception) return the list of
// results (or raise the first raised exception).
public fun async-all( actions : list<() -> <async|e> a> ) : <async|e> list<a> {
  async-all-prim( actions.vector ).list;
}


extern async-all-prim( actions : vector<() -> <async|e> a> ) : <async|e> vector<a>  {
  js inline "$asyncAll(actions,_k)"
}

extern async-callback0( action: () -> e (), haserr : bool, nodup : bool ) : (() -> e ()) {
  js "$asyncCallback"
}

public fun unsafe-async-yield-err( action : (cb : a -> <async|e> ()) -> <async|e> () ) : <async|e> a {
  unsafe-async-yield( action, True ) 
}

public extern unsafe-async-yield( action : (cb : a -> <async|e> ()) -> <async|e> (), haserr : bool = False ) : <async|e> a {
  js inline "(#1)( $asyncCallback(_k,#2), $std_core.async_yield)"
}

// Wait (asynchronously) for `secs` seconds. 
// Use `wait(0)` to yield to other asynchronous operations.
public fun wait( secs : double ) : <async,exn> () {
  wait(secs.duration)
}

// Wait (asynchronously) for `secs` seconds duration. 
// Use `wait(0)` to yield to other asynchronous operations.
public fun wait( secs : duration ) : <async,exn> () {
  unsafe-async-yield( fun(cb) { 
    val msecs = secs.milli-seconds.int32
    val h = set-timeout( cb, msecs)
    if (msecs > 100.int32) { 
      unsafe-register-cancel( cb, fun(){ cancel-timeout(h) } )
    }
  })
}

type timeout-handle

extern set-timeout( action : (()) -> e (), ms : int32 ) : exn timeout-handle {
  js "setTimeout"
}

extern cancel-timeout( h : timeout-handle ) : exn () {
  js "clearTimeout"
}


// Execute any asynchronous actions inside `action` synchronously, ie. 
// no other actions are interleaved.
public fun synchronous( action : () -> <async|e> a ) : e a {
  start-synchronous()
  on-exit(end-synchronous){ unsafe-total(action) }
}

extern start-synchronous() : () {
  js "$asyncStartSynchronous"
}

extern end-synchronous() : () {
  js "$asyncEndSynchronous"
}

