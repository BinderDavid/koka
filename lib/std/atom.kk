/*---------------------------------------------------------------------------
  Copyright 2012 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

/* This module provides a perfect hash for strings.

   These `:atom`s can be compared very efficiently. If you compare only
   seldomly, a straight string compare will probably be faster since 
   creating an atom takes some initial work.
*/
module std/atom

import std/vector

external include {
  cs file "atom-inline.cs"
  js file "atom-inline.js"
}

// An `atom` corresponding to the empty string
public val atom-null = atom("")

// An atom is an abstract representation of a string that can be compared quickly
abstract struct atom( value: int )  

// Create an `:atom`
extern atom-create : ( s : string ) -> int {
  cs "Atom.Create"
  js "$atomCreate"
}

// Convert an `:atom` back to a `:string`
extern atom-to-string : ( x : int ) -> string {
  cs "Atom.ToString"
  js "$atomToString"
}

// Create an `:atom`
public fun atom( s : string ) : atom {
  atom-create(s).Atom
}

// Convert an `:atom` back to a `:string`
public fun string( a : atom ) : string {
  atom-to-string(a.value)
}

// Compare two atoms for equality
public fun (==)(x : atom, y : atom) : bool  -> (x.value == y.value)

// Compare if two atoms are not equal.
public fun (!=)(x : atom, y : atom) : bool  -> (x.value != y.value)


extern atom-matcher : ( xs : vector<string>, ignoreCase: int ) -> any
{
  cs "Atom.CreateMatcher"
  js "$atomCreateMatcher"
}

extern atom-match : ( m : any, s : string, ignoreCase: int ) -> int
{
  cs "Atom.Match"
  js "$atomMatch"
}

// Create an efficient string matcher which given a list of strings returns a matching function.
public fun matcher( xs : list<string>, ignoreCase: bool = False ) : (( s : string ) -> bool)
{
  matcher(vector(xs), ignoreCase)
}

// Create an efficient string matcher which given a vector of strings returns a matching function.
public fun matcher( xs : vector<string>, ignoreCase: bool = False ) : (( s : string ) -> bool)
{
  val m = atom-matcher( xs, ignoreCase.int );
  return fun(s) { return bool(atom-match(m,s,int(ignoreCase))); }
}
