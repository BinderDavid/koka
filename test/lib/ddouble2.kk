/*
  seconds of days: 9 223 372 035 908 035 200
    + 55808
 =  9 223 372 035 908 091 008
*/

import std/num/ddouble

fun main() {
  val i = 9223372035908091008
  val x : ddouble = i.ddouble
  val y = ddouble(9223372035908091000.0) + ddouble(8.0)
  val z = ddouble(9223372035908090880.0) + ddouble(128.0)
  println( x.show-sum(20) )
  println( y.show-sum )
  println( z.show-sum )

  println( i.show )
  println( x.show-fixed )
  println( y.int.show )
  println( z.int.show )
}