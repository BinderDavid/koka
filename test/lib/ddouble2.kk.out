e : ok: 2.7182818284590452353602874713527
pi-mach : ok: 3.141592653589793238462643383279
borwein : ok: 3.14159265358979323846264338328
pi-sb : ok: 3.141592653589793238462643383279
log2 : ok: 0.6931471805599453094172321214582
exp1 : ok: 7.3890560989306502272304274605750
exp2 : ok: 7.389056098930650227230427460575
exp3 : ok: 7.389056098930650227230427460575
sin(6.283185307179586476925286766559): ok: 1.078520768856852077711347647848e-32, dt: 1.07852e-32
cos(6.283185307179586476925286766559): ok: 1, dt: 1.54074e-32
dsin0 : ok: 0, dt: 2.44929e-16
dcos0 : ok: 1, dt: 0
sin(-6.283185307179586476925286766559): ok: -1.078520768856852077711347647848e-32, dt: 1.07852e-32
cos(-6.283185307179586476925286766559): ok: 1, dt: 1.54074e-32
dsin0 : ok: 0, dt: 2.44929e-16
dcos0 : ok: 1, dt: 0
sin(3.1415926535897932384626433832795): ok: -6.162975822039154729779129416272e-33, dt: 6.16298e-33
cos(3.1415926535897932384626433832795): ok: -1, dt: 6.16298e-33
dsin1 : ok: 0, dt: 1.22465e-16
dcos1 : ok: -1, dt: 0
sin(-3.1415926535897932384626433832795): ok: 6.162975822039154729779129416272e-33, dt: 6.16298e-33
cos(-3.1415926535897932384626433832795): ok: -1, dt: 6.16298e-33
dsin1 : ok: 0, dt: 1.22465e-16
dcos1 : ok: -1, dt: 0
sin(1.5707963267948966192313216916398): ok: 1, dt: 2.54223e-32
cos(1.5707963267948966192313216916398): ok: -3.081487911019577364889564708136e-33, dt: 3.08149e-33
dsin2 : ok: 1, dt: 0
dcos2 : ok: 0, dt: 6.12323e-17
sin(-1.5707963267948966192313216916398): ok: -1, dt: 2.54223e-32
cos(-1.5707963267948966192313216916398): ok: -3.081487911019577364889564708136e-33, dt: 3.08149e-33
dsin2 : ok: -1, dt: 0
dcos2 : ok: 0, dt: 6.12323e-17
sin(2.0943951023931954923084289221863): ok: 0.8660254037844386467637231707529, dt: 6.77927e-32
cos(2.0943951023931954923084289221863): ok: -0.5, dt: 6.31705e-32
dsin3 : ok: 0.866025403784438646763723170753, dt: 5.01754e-17
dcos3 : ok: -0.5, dt: 2.22045e-16
sin(-2.0943951023931954923084289221863): ok: -0.8660254037844386467637231707529, dt: 6.77927e-32
cos(-2.0943951023931954923084289221863): ok: -0.5, dt: 6.31705e-32
dsin3 : ok: -0.866025403784438646763723170753, dt: 5.01754e-17
dcos3 : ok: -0.5, dt: 2.22045e-16
sin(1.2566370614359172953850573533118): ok: 0.9510565162951535721164393333794, dt: 1.23260e-32
cos(1.2566370614359172953850573533118): ok: 0.3090169943749474241022934171828, dt: 2.77334e-32
dsin4 : ok: 0.9510565162951535721164393333794, dt: 4.09345e-17
dcos4 : ok: 0.3090169943749474241022934171829, dt: 2.71606e-17
sin(-1.2566370614359172953850573533118): ok: -0.9510565162951535721164393333794, dt: 1.23260e-32
cos(-1.2566370614359172953850573533118): ok: 0.3090169943749474241022934171828, dt: 2.77334e-32
dsin4 : ok: -0.9510565162951535721164393333794, dt: 4.09345e-17
dcos4 : ok: 0.3090169943749474241022934171829, dt: 2.71606e-17
sin(1.0471975511965977461542144610932): ok: 0.866025403784438646763723170753, dt: 6.16298e-33
cos(1.0471975511965977461542144610932): ok: 0.5, dt: 1.69482e-32
dsin5 : ok: 0.866025403784438646763723170753, dt: 6.08469e-17
dcos5 : ok: 0.5, dt: 1.11022e-16
sin(-1.0471975511965977461542144610932): ok: -0.866025403784438646763723170753, dt: 6.16298e-33
cos(-1.0471975511965977461542144610932): ok: 0.5, dt: 1.69482e-32
dsin5 : ok: -0.866025403784438646763723170753, dt: 6.08469e-17
dcos5 : ok: 0.5, dt: 1.11022e-16
sin(0.897597901025655210989326680937): ok: 0.7818314824680298087084445266741, dt: 3.92890e-32
cos(0.897597901025655210989326680937): ok: 0.6234898018587335305250048840042, dt: 9.86076e-32
dsin6 : ok: 0.7818314824680298087084445266741, dt: 5.07432e-18
dcos6 : ok: 0.6234898018587335305250048840043, dt: 6.38613e-17
sin(-0.897597901025655210989326680937): ok: -0.7818314824680298087084445266741, dt: 3.92890e-32
cos(-0.897597901025655210989326680937): ok: 0.6234898018587335305250048840042, dt: 9.86076e-32
dsin6 : ok: -0.781831482468029808708444526674, dt: 5.07432e-18
dcos6 : ok: 0.6234898018587335305250048840043, dt: 6.38613e-17
sin(0.7853981633974483096156608458199): ok: 0.7071067811865475244008443621049, dt: 6.16298e-33
cos(0.7853981633974483096156608458199): ok: 0.7071067811865475244008443621049, dt: 1.84889e-32
dsin7 : ok: 0.7071067811865475244008443621049, dt: 6.26858e-17
dcos7 : ok: 0.7071067811865475244008443621049, dt: 4.83365e-17
sin(-0.7853981633974483096156608458199): ok: -0.7071067811865475244008443621049, dt: 6.16298e-33
cos(-0.7853981633974483096156608458199): ok: 0.7071067811865475244008443621049, dt: 1.84889e-32
dsin7 : ok: -0.7071067811865475244008443621048, dt: 6.26858e-17
dcos7 : ok: 0.7071067811865475244008443621049, dt: 4.83365e-17
sin(0.6981317007977318307694763073955): ok: 0.6427876096865393263226434099072, dt: 0
cos(0.6981317007977318307694763073955): ok: 0.7660444431189780352023926505555, dt: 9.24446e-33
dsin8 : ok: 0.6427876096865393263226434099072, dt: 7.44262e-17
dcos8 : ok: 0.7660444431189780352023926505555, dt: 2.17507e-17
sin(-0.6981317007977318307694763073954): ok: -0.6427876096865393263226434099072, dt: 0
cos(-0.6981317007977318307694763073954): ok: 0.7660444431189780352023926505555, dt: 9.24446e-33
dsin8 : ok: -0.6427876096865393263226434099072, dt: 7.44262e-17
dcos8 : ok: 0.7660444431189780352023926505555, dt: 8.92716e-17
sin(0.6283185307179586476925286766559): ok: 0.5877852522924731291687059546391, dt: 5.54668e-32
cos(0.6283185307179586476925286766559): ok: 0.8090169943749474241022934171828, dt: 1.54074e-32
dsin9 : ok: 0.5877852522924731291687059546391, dt: 7.93475e-18
dcos9 : ok: 0.8090169943749474241022934171828, dt: 2.71606e-17
sin(-0.6283185307179586476925286766559): ok: -0.5877852522924731291687059546391, dt: 5.54668e-32
cos(-0.6283185307179586476925286766559): ok: 0.8090169943749474241022934171828, dt: 1.54074e-32
dsin9 : ok: -0.587785252292473129168705954639, dt: 7.93475e-18
dcos9 : ok: 0.8090169943749474241022934171828, dt: 2.71606e-17
sin(0.0062831853071795864769252867665): ok: 0.0062831439655589512497353101579, dt: 8.90743e-34
cos(0.0062831853071795864769252867665): ok: 0.9999802608561371298486736604302, dt: 3.08149e-33
dsin10 : ok: 0.0062831439655589512497353101579, dt: 1.65475e-19
dcos10 : ok: 0.9999802608561371298486736604302, dt: 1.45158e-17
sin(-0.0062831853071795864769252867666): ok: -0.006283143965558951249735310158, dt: 8.90743e-34
cos(-0.0062831853071795864769252867666): ok: 0.9999802608561371298486736604302, dt: 3.08149e-33
dsin10 : ok: -0.006283143965558951249735310158, dt: 1.65475e-19
dcos10 : ok: 0.9999802608561371298486736604302, dt: 1.45158e-17
sin(6.283185307179586476925286766559e-10): ok: 6.283185307179586476511869744151e-10, dt: 5.73972e-42
cos(6.283185307179586476925286766559e-10): ok: 0.9999999999999999998026079119782, dt: 0
dsin11 : ok: 6.283185307179586476511869744151e-10, dt: 2.80351e-26
dcos11 : ok: 0.9999999999999999998026079119782, dt: 1.97392e-19
sin(-6.283185307179586476925286766559e-10): ok: -6.283185307179586476511869744151e-10, dt: 5.73972e-42
cos(-6.283185307179586476925286766559e-10): ok: 0.9999999999999999998026079119782, dt: 0
dsin11 : ok: -6.283185307179586476511869744151e-10, dt: 2.80351e-26
dcos11 : ok: 0.9999999999999999998026079119782, dt: 1.97392e-19